include/boost/geometry/algorithms/detail/overlay/traversal.hpp

```diff
@@ -591,14 +576,14 @@ public :
 
         if BOOST_GEOMETRY_CONSTEXPR (OverlayType != overlay_dissolve)
         {
-            if (op.enriched.count_left != 0 || op.enriched.count_right == 0)
+            if ((op.enriched.count_left != 0 || op.enriched.count_right == 0) && cinfo.spike_count > 0)
             {
                 // Check counts: in some cases interior rings might be generated with
                 // polygons on both sides. For dissolve it can be anything.
 
                 // If this forms a spike, going to/from the cluster point in the same
                 // (opposite) direction, it can still be used.
-                return cinfo.spike_count > 0 ? 1 : 0;
+                return 1;
```

Comment:
Returning `0` was a bug, it should only have returned 1.
So that is fixed and is necessary to avoid regressions with this PR